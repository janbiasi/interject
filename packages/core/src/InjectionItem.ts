import { LookupFlags } from './Flags';

export interface IInjection {
	token: any;
	factory: Function;
	useNew: boolean;
	dependencies: IInjectionDependency[];
	value: any;
	useExisting: any;
}

export interface IInjectionDependency {
	token: any;
	options: LookupFlags;
}
