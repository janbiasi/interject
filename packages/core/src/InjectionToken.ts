import { ProviderScope } from './Scope';
import { IInjectable } from './Injectable';

interface InjectionTokenOptions<T> {
	scope: ProviderScope;
	factory: () => T;
}

const INJECTION_TOKEN_DEFAULT_OPTIONS = {
	scope: 'root',
	factory: () => undefined,
};

export class InjectionToken<T> {
	readonly metadataName = 'InjectionToken';
	readonly injectable: IInjectable<T>;

	constructor(private readonly _id: string, options: InjectionTokenOptions<T> = INJECTION_TOKEN_DEFAULT_OPTIONS) {
		this.injectable = {
			scope: options.scope,
			factory: options.factory,
			value: undefined,
		};
	}

	toString() {
		const scope = this.injectable.scope;
		const displayPossibleScope = scope && scope !== 'root' ? `@${scope}` : '';

		return `InjectionToken${displayPossibleScope} ${this._id}`;
	}
}
