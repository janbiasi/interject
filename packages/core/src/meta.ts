import 'reflect-metadata';

export const enum TypeDeclaration {
	Component = 'Component',
}

export const metadataKeys = {
	injectable: 'custom:injectable',
	token: 'custom:token',
	provider: 'custom:provider',
	injector: 'custom:injector',
	bootstrap: 'custom:bootstrap',
};

export function getMetadata<T extends any>(metadataKey: any, target: any): T {
	return Reflect.getMetadata(metadataKey, target);
}

export function defineMetadata(metadataKey: any, metadataValue: any, target: any): void {
	Reflect.defineMetadata(metadataKey, metadataValue, target);
}
