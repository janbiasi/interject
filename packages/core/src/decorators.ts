import { IConstructable } from './Constructable';
import { IClassProvider, Provider } from './Provider';
import { InjectionToken } from './InjectionToken';
import { ProviderScope } from './Scope';
import { IInjectableType, IInjectable } from './Injectable';
import { defineMetadata, metadataKeys, getMetadata } from './meta';
import { Injector, CommonInjector } from '.';
import { stringify } from './utils';

export interface IInjectableOptions {
	scope?: ProviderScope;
	requires?: Provider[];
}

const INJECTABLE_DEFAULT_OPTS = {
	scope: 'root',
	requires: [],
};

export function Injectable({ scope, requires }: IInjectableOptions = INJECTABLE_DEFAULT_OPTS) {
	return <T extends { new (...args: any[]): any }>(Factory: T) => {
		const token = new InjectionToken<T>(Factory.name || stringify(Factory));
		const injectable = <IInjectable<T>>{
			scope,
			requires,
			useNew: true,
			factory: Factory as any,
			value: undefined,
		};

		defineMetadata(metadataKeys.token, token, Factory);
		defineMetadata(metadataKeys.injectable, injectable, Factory);
	};
}

export interface IModuleOptions {
	providers?: Array<IConstructable<any> | Provider>;
	parent?: CommonInjector
}

const MODULE_DEFAULT_OPTS = {
	providers: [],
};

/** @internal */
export function internalModuleDecoratorFactory<T extends IConstructable<any>>(
	{ providers, parent }: IModuleOptions,
	Factory: T
): CommonInjector {
	const processedProviders: Provider[] = providers.map(
		(provider: any): Provider => {
			if (typeof provider === 'function') {
				return {
					provide: getMetadata(metadataKeys.token, provider),
					useClass: provider,
				};
			}

			return provider as Provider;
		}
	);

	// create module root injector without parent
	// TODO: let module inherit from other module for correct resolving of inherited deps
	const injector = CommonInjector.create({
		providers: processedProviders,
		parent: parent || CommonInjector.NULL,
		name: Factory.name,
	});

	defineMetadata(metadataKeys.injector, injector, Factory);
	defineMetadata(
		metadataKeys.bootstrap,
		// the tokens in correct order for bootstrap of the module
		processedProviders.map((provider: Provider) => provider.provide),
		Factory
	);

	return injector;
}

export function Module({ providers, parent }: IModuleOptions = MODULE_DEFAULT_OPTS) {
	return <T extends { new (...args: any[]): any }>(Factory: T) => {
		internalModuleDecoratorFactory({
			providers,
			parent
		}, Factory);
	};
}
