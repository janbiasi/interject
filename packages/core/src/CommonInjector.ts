import { IConstructable } from './Constructable';
import { Provider } from './Provider';
import { THROW_IF_NOT_FOUND } from './utils';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags } from './Flags';
import { IInjectable } from './Injectable';
import { Injector } from './Injector';

// runtime
import { setCurrentInjector, getCurrentInjector } from './runtime';

// core providers
import VersionProvider from './platform/providers/Version';

export interface IInjectorType<T> extends IConstructable<T> {
	injector: IInjector<T>;
}

export interface IInjectorTypeWithProviders<T> {
	pointer: IInjectorType<T>;
	providers?: Provider[];
}

export interface IInjector<T> {
	factory: () => T;
	providers: Provider[];
	imports: (IInjectorType<any> | IInjectorTypeWithProviders<any>)[];
}

export interface ICommonInjectorCreateOptions {
	providers: Provider[];
	parent?: CommonInjector;
	name?: string;
}

export class NullInjector implements CommonInjector {
	get(token: any, notFoundValue: any = THROW_IF_NOT_FOUND): any {
		if (notFoundValue === THROW_IF_NOT_FOUND) {
			throw new Error(`NullInjectorError: No provider for ${token}!`);
		}

		return notFoundValue;
	}
}

export abstract class CommonInjector {
	static NULL: CommonInjector = new NullInjector();
	static THROW_IF_NOT_FOUND = THROW_IF_NOT_FOUND;

	abstract get<T>(token: IConstructable<T> | InjectionToken<T>, notFoundValue?: T, flags?: InjcetionFlags): T;

	static create(options: ICommonInjectorCreateOptions): CommonInjector {
		// add core providers
		options.providers.push(VersionProvider);

		return new Injector(options.providers, options.parent, options.name || null);
	}

	static injectable = <IInjectable<any>>{
		providedIn: 'any',
		factory: () => undefined as any,
		scope: 'root',
		value: undefined,
	};

	static toString() {
		return `CommonInjector`;
	}

	static get currentInjector(): CommonInjector {
		return getCurrentInjector();
	}
}
