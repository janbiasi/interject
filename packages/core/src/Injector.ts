import { CommonInjector } from './CommonInjector';
import { Provider } from './Provider';
import { IInjection } from './InjectionItem';
import { EMPTY, CIRCULAR, stringify } from './utils';
import { InjectionToken } from './InjectionToken';
import { IConstructable } from './Constructable';
import { InjcetionFlags } from './Flags';
import { resolveToken, resolveInjectableProvider } from './resolve';
import VersionProvider, { VersionProviderToken } from './platform/Version';

export class Injector implements CommonInjector {
	readonly parent: CommonInjector;
	readonly source: string | null;

	private _injections: Map<any, IInjection>;

	constructor(providers: Provider[], parent: CommonInjector = CommonInjector.NULL, source: string | null = null) {
		this.parent = parent;
		this.source = source;

		// storage for all injectables
		const injections = (this._injections = new Map<any, IInjection>());

		// set an injection point to the common injector
		injections.set(CommonInjector, <IInjection>{
			token: CommonInjector,
			factory: CIRCULAR,
			dependencies: EMPTY,
			value: this,
			useNew: false,
		});

		// declare version provider
		// TODO: Outsource platform providers
		injections.set(VersionProviderToken, <IInjection>{
			token: VersionProviderToken,
			factory: CIRCULAR,
			dependencies: EMPTY,
			value: VersionProvider.useValue,
			useNew: false,
		});

		// process all providers
		providers.forEach((provider: Provider) => resolveInjectableProvider(provider, injections));
	}

	get<T>(token: InjectionToken<any> | IConstructable<any>, notFoundValue?: T, flags?: InjcetionFlags): T {
		const injection = this._injections.get(token);

		// resolve the token and its dependencies
		return resolveToken(token, injection, this._injections, this.parent, notFoundValue, flags);
	}

	toString() {
		const tokens: string[] = [];
		this._injections.forEach((key, token) => tokens.push(stringify(token)));

		return `Injector {Â ${tokens.join(', ')} }`;
	}
}
