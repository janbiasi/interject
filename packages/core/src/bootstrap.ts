import { IConstructable } from './Constructable';
import { getMetadata, metadataKeys } from './meta';
import { CommonInjector } from './CommonInjector';
import { Provide } from './Provider';
import { InjectionToken } from './InjectionToken';

export function bootstrap<T>(ModuleFactory: IConstructable<T>): T {
	const injector: CommonInjector = getMetadata(metadataKeys.injector, ModuleFactory);
	const tokens: (InjectionToken<any> | IConstructable<any>)[] = getMetadata(metadataKeys.bootstrap, ModuleFactory);
	const providers = tokens.map((token: Provide) => injector.get(token));
	const instance = new ModuleFactory(...providers);

	return instance;
}
