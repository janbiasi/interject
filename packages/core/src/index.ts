import { CommonInjector } from './CommonInjector';
import { Injector } from './Injector';
import { InjectionToken } from './InjectionToken';
import { InjcetionFlags, LookupFlags } from './Flags';
import { Provider } from './Provider';

// platform providers
import VersionProvider from './platform/Version';

export { Injector, CommonInjector, InjectionToken, InjcetionFlags, LookupFlags, Provider, VersionProvider };
