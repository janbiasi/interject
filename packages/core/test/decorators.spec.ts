import { Injectable, Module } from '../src/decorators';
import { InjectionToken } from '../src/InjectionToken';
import { bootstrap } from '../src/bootstrap';

describe('decorators', () => {
	it('should return an inject providers correctly', () => {
		@Injectable({
			scope: 'root',
		})
		class TestService {
			public get value() {
				return 10;
			}
		}

		const VALUE_TOKEN = new InjectionToken<number>('value');

		@Module({
			providers: [TestService, { provide: VALUE_TOKEN, useValue: 20 }],
		})
		class AppModule {
			constructor(private _testService: TestService, private _value: number) {}

			public get testService() {
				return this._testService;
			}

			public get value() {
				return this._value;
			}
		}
		const moduleInstance = bootstrap(AppModule);

		expect(moduleInstance).toMatchSnapshot();
		expect(moduleInstance.testService.value).toEqual(10);
		expect(moduleInstance.value).toEqual(20);
	});
});
