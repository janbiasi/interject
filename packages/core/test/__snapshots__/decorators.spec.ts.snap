// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators should return an inject providers correctly 1`] = `
AppModule {
  "_testService": TestService {},
  "_value": 20,
}
`;
